service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,271 @@
+        for (GiftCertificate giftCertificate : giftCertificateList) {
+            giftCertificateDtoList.add(buildGiftCertificateDto(giftCertificate));
+        }
+        return giftCertificateDtoList;
+    }
+
+    @Override
+    public List<GiftCertificateDto> getAllWithSortingAndFiltering(List<String> sortColumns,
+                                                                  List<String> orderType,
+                                                                  List<String> filterBy) {
+        List<GiftCertificateDto> giftCertificateDtoList = new ArrayList<>();
+        if (sortColumns == null) {
+            sortColumns = Collections.emptyList();
+        }
+        if (orderType == null) {
+            orderType = Collections.emptyList();
+        }
+        if (filterBy == null) {
+            filterBy = Collections.emptyList();
+        }
+        List<GiftCertificate> giftCertificateList = giftCertificateDao.getAllWithSortingAndFiltering(sortColumns, orderType, filterBy);
+        for (GiftCertificate giftCertificate : giftCertificateList) {
+            giftCertificateDtoList.add(buildGiftCertificateDto(giftCertificate));
+        }
+        return giftCertificateDtoList;
+    }
+
+    @Transactional
+    @Override
+    public GiftCertificateDto updateById(Long giftCertificateId, GiftCertificateDto giftCertificateDto) {
+        GiftCertificate giftCertificate = giftCertificateDto.getGiftCertificate();
+        List<Tag> tagList = giftCertificateDto.getCertificateTags();
+        if (tagList == null || giftCertificate == null) {
+            throw new InvalidEntityException("message.cantUpdateGiftCertificateById");
+        }
+        isPresentGiftCertificate(giftCertificateId);
+        giftCertificateDao.updateById(giftCertificateId, updateInfo(giftCertificate));
+        updateCertificateTags(tagList, giftCertificateId);
+        return buildGiftCertificateDto(giftCertificateDao.getById(giftCertificateId).orElseThrow(() -> new NoSuchEntityException("message.notFound")));
+    }
+
+    private Map<String, Object> updateInfo(GiftCertificate giftCertificate) {
+        Map<String, Object> updateInfo = new HashMap<>();
+        isValidNameForUpdate(giftCertificate, updateInfo, giftCertificateValidator);
+        isValidDescriptionForUpdate(giftCertificate, updateInfo, giftCertificateValidator);
+        isPriceValidForUpdate(giftCertificate, updateInfo, giftCertificateValidator);
+        isDurationValidForUpdate(giftCertificate, updateInfo, giftCertificateValidator);
+        return updateInfo;
+    }
+
+    private String getGiftCertificateName(GiftCertificate giftCertificate) {
+        String giftCertificateName = giftCertificate.getName();
+        boolean isCertificateExist = giftCertificateDao.getByName(giftCertificateName).isPresent();
+        if (isCertificateExist) {
+            throw new DuplicateEntityException("message.giftCertificateExists");
```

Comment:
Move it validator

Reply:
can i do it like this?
private String getGiftCertificateName(GiftCertificate giftCertificate) {
        String giftCertificateName = giftCertificate.getName();
        giftCertificateDao.getByName(giftCertificateName).orElseThrow(()-> new DuplicateEntityException("message.giftCertificateExists"));
        return giftCertificateName;
    }

Reply:
yes